### ISSUE TITLE
new hosts failing to register with SGE master due to cfn-init failure

### ISSUE THREAD

**Issue by keien** [aws/aws-parallelcluster#866]
**Environment:**
 - AWS ParallelCluster / CfnCluster version: 2.1.1
 - OS: Centos 7.6
 - Scheduler: SGE
 - Master instance type: t2
 - Compute instance type: m5/c5/others

**Bug description and how to reproduce:**

Since about 10AM this morning, all new nodes on our clusters are failing to register themselves with the SGE master resulting in numerous orphaned nodes.

They are all failing the same way with the following message left in `/var/log/messages`:

```
Feb 18 12:50:35 ip-172-31-128-62 cloud-init: Traceback (most recent call last):
Feb 18 12:50:35 ip-172-31-128-62 cloud-init: File "/usr/lib64/python2.7/logging/__init__.py", line 874, in emit
Feb 18 12:50:35 ip-172-31-128-62 cloud-init: stream.write(fs % msg.encode("UTF-8"))
Feb 18 12:50:35 ip-172-31-128-62 cloud-init: UnicodeDecodeError: 'ascii' codec can't decode byte 0xe2 in position 61: ordinal not in range(128)
Feb 18 12:50:35 ip-172-31-128-62 cloud-init: Logged from file util.py, line 476
Feb 18 12:50:36 ip-172-31-128-62 cloud-init: + cfn-signal --exit-code=0 '--reason=MasterServer setup complete' --stack=parallelcluster-m5-4xlarge-ami-02860d917774b37b8-spot --resource=ComputeFleet --region=us-west-2
Feb 18 12:50:36 ip-172-31-128-62 cloud-init: ValidationError: Stack arn:aws:cloudformation:us-west-2:171496337684:stack/parallelcluster-m5-4xlarge-ami-02860d917774b37b8-spot/c37e25c0-30bc-11e9-81f2-064b816f3a4a is in CREATE_COMPLETE state and cannot be signaled
Feb 18 12:50:36 ip-172-31-128-62 cloud-init: 2019-02-18 20:50:36,520 - util.py[WARNING]: Failed running /var/lib/cloud/instance/scripts/part-002 [1]
Feb 18 12:50:36 ip-172-31-128-62 cloud-init: /usr/bin/yum
Feb 18 12:50:36 ip-172-31-128-62 cloud-init: which: no apt-get in (/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin)
Feb 18 12:50:36 ip-172-31-128-62 cloud-init: 2019-02-18 20:50:36,530 - util.py[WARNING]: Failed running /var/lib/cloud/instance/scripts/runcmd [1]
Feb 18 12:50:36 ip-172-31-128-62 cloud-init: 2019-02-18 20:50:36,530 - cc_scripts_user.py[WARNING]: Failed to run module scripts-user (scripts in /var/lib/cloud/instance/scripts)
Feb 18 12:50:36 ip-172-31-128-62 cloud-init: 2019-02-18 20:50:36,531 - util.py[WARNING]: Running module scripts-user (<module 'cloudinit.config.cc_scripts_user' from '/usr/lib/python2.7/site-packages/cloudinit/config/cc_scripts_user.pyc'>) failed
Feb 18 12:50:36 ip-172-31-128-62 cloud-init: 2019-02-18 20:50:36,539 - util.py[WARNING]: restorecon failed on /home/centos/.ssh,True maybe badness? [Errno 95] Operation not supported
```

**Comment by keien** [comment]
seems the above issues were a red herring; the real issue appears to be that sqswatcher on the master was stuck and the cluster SQS queues weren't draining - restarting sqswatcher seems to have fixed the issue but it's unclear why it got stuck in the first place, or why this happened on multiple clusters

**Comment by sean-smith** [comment]
Can you post the sqswatcher logs? Grep for error in them and post what you find.

On Feb 18, 2019, at 4:51 PM, Keien Ohta <notifications@github.com<mailto:notifications@github.com>> wrote:


seems the above issues were a red herring; the real issue appears to be that sqswatcher on the master was stuck and the cluster SQS queues weren't draining - restarting sqswatcher seems to have fixed the issue but it's unclear why it got stuck in the first place, or why this happened on multiple clusters

**Comment by keien** [comment]
posted below, but not that relevant as this log hasn't been updated since 2 days ago...

```
[ip-172-31-128-112]root:sqswatcher [17:21]
$ grep error /var/log/sqswatcher
2019-02-14 20:19:04,569 ERROR [sge:addHost] Socket error: [Errno 110] Connection timed out
2019-02-14 20:19:17,581 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.128
2019-02-14 20:19:31,601 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.128
2019-02-14 20:30:21,815 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.103
2019-02-14 20:30:34,829 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.103
2019-02-14 20:30:48,849 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.103
2019-02-14 23:59:46,809 ERROR [sge:addHost] Socket error: [Errno 110] Connection timed out
2019-02-14 23:59:59,829 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.150
2019-02-15 00:00:13,837 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.150
2019-02-15 02:47:01,303 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.159
2019-02-15 02:47:14,321 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.159
2019-02-15 02:47:28,337 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.159
2019-02-15 03:00:19,705 ERROR [sge:addHost] Socket error: [Errno 110] Connection timed out
2019-02-15 03:00:32,725 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.143
2019-02-15 03:00:46,745 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.143
2019-02-15 03:10:00,735 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.241
2019-02-15 03:12:18,041 ERROR [sge:addHost] Socket error: [Errno 110] Connection timed out
2019-02-15 03:12:32,053 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.241
2019-02-15 03:24:23,289 ERROR [sge:addHost] Socket error: [Errno 110] Connection timed out
2019-02-15 03:24:36,309 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.81
2019-02-15 03:24:50,327 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.81
2019-02-15 04:47:34,071 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.98
2019-02-15 04:47:47,089 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.98
2019-02-15 04:48:01,109 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.98
2019-02-15 05:00:22,521 ERROR [sge:addHost] Socket error: [Errno 110] Connection timed out
2019-02-15 05:00:35,533 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.56
2019-02-15 05:00:49,551 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.56
2019-02-15 05:24:43,001 ERROR [sge:addHost] Socket error: [Errno 110] Connection timed out
2019-02-15 05:24:56,017 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.178
2019-02-15 05:25:10,037 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.178
2019-02-15 08:23:39,799 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.39
2019-02-15 08:23:52,817 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.39
2019-02-15 08:24:06,837 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.39
2019-02-15 09:12:08,695 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.164
2019-02-15 09:12:21,715 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.164
2019-02-15 09:12:35,733 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.164
2019-02-15 11:15:25,433 ERROR [sge:addHost] Socket error: [Errno 110] Connection timed out
2019-02-15 11:15:38,453 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.198
2019-02-15 11:15:52,473 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.198
2019-02-15 11:39:45,785 ERROR [sge:addHost] Socket error: [Errno 110] Connection timed out
2019-02-15 11:39:58,805 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.39
2019-02-15 11:40:12,825 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.39
2019-02-15 12:27:13,783 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.216
2019-02-15 12:27:26,803 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.216
2019-02-15 12:27:40,821 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.216
2019-02-15 12:40:02,169 ERROR [sge:addHost] Socket error: [Errno 110] Connection timed out
2019-02-15 12:40:15,179 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.104
2019-02-15 12:40:29,195 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.104
2019-02-15 13:38:57,719 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.152
2019-02-15 13:39:10,737 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.152
2019-02-15 13:39:24,747 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.152
2019-02-15 16:20:30,585 ERROR [sge:addHost] Socket error: [Errno 110] Connection timed out
2019-02-15 16:20:43,597 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.167
2019-02-15 16:20:57,609 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.167
2019-02-15 19:54:06,327 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.92
2019-02-15 19:54:19,349 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.92
2019-02-15 19:54:33,365 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.92
2019-02-15 23:06:45,175 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.13
2019-02-15 23:06:58,193 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.13
2019-02-15 23:07:12,215 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.13
2019-02-15 23:31:35,673 ERROR [sge:addHost] Socket error: [Errno 110] Connection timed out
2019-02-15 23:31:48,693 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.141
2019-02-15 23:32:02,701 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.141
2019-02-16 02:50:00,441 ERROR [sge:addHost] Socket error: [Errno 110] Connection timed out
2019-02-16 02:50:13,461 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.162
2019-02-16 02:50:27,469 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.162
2019-02-16 05:38:09,468 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.161
2019-02-16 05:40:26,745 ERROR [sge:addHost] Socket error: [Errno 110] Connection timed out
2019-02-16 05:40:40,765 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.161
2019-02-16 13:46:57,527 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.120
2019-02-16 13:47:10,545 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.120
2019-02-16 13:47:24,565 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.120
2019-02-16 21:54:53,433 ERROR [sge:addHost] Socket error: [Errno 110] Connection timed out
2019-02-16 21:55:06,449 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.85
2019-02-16 21:55:20,469 ERROR [sge:addHost] Socket error: [Errno None] Unable to connect to port 22 on 172.31.128.85
```

**Comment by keien** [comment]
gdb on sqswatcher shows
```
(gdb) py-bt
#4 Frame 0x34e7320, for file /usr/lib/python2.7/site-packages/sqswatcher/plugins/sge.py, line 95, in addHost (hostname='ip-172-31-128-238', cluster_user='centos', slots=u'36', command="sudo sh -c 'cd /opt/sge && /opt/sge/inst_sge -noremote -x -auto /opt/parallelcluster/templates/sge/sge_inst.conf'", qconf_Ae_template='hostname              %s\nload_scaling          NONE\ncomplex_values        NONE\nuser_lists            NONE\nxuser_lists           NONE\nprojects              NONE\nxprojects             NONE\nusage_scaling         NONE\nreport_variables      NONE\n', t=<_TemporaryFileWrapper(close_called=True, fileno=<built-in method fileno of file object at remote 0x7f99266b78a0>, name='/tmp/tmpr8MixS', file=<file at remote 0x7f99266b78a0>, delete=True) at remote 0x7f992644cea8>, temp_template=<file at remote 0x7f9924177ae0>, ssh=<SSHClient(_system_host_keys=<HostKeys(_entries=[]) at remote 0x7f992668ad50>, _policy=<AutoAddPolicy at remote 0x7f9926584590>, _log_channel=None, _transport=<Transport(default_window_...(truncated)
    time.sleep(1)
#8 Frame 0x2f932c0, for file /usr/lib/python2.7/site-packages/sqswatcher/sqswatcher.py, line 155, in pollQueue (scheduler='sge', q=<sqs.Queue(meta=<ResourceMeta(service_name='sqs', identifiers=[u'url'], data=None, resource_model=<ResourceModel(_resource_defs=<OrderedDict(_OrderedDict__root=[[[[...], [...], u'Message'], [...], u'Queue'], [...], None], _OrderedDict__map={u'Queue': [...], u'Message': [...]}) at remote 0x7f9927678868>, shape=u'GetQueueAttributesResult', _definition=<OrderedDict(_OrderedDict__root=[[[[[[[[...], [...], u'identifiers'], [...], u'shape'], [...], u'load'], [...], u'actions'], [...], u'has'], [...], u'hasMany'], [...], None], _OrderedDict__map={u'load': [...], u'identifiers': [...], u'hasMany': [...], u'actions': [...], u'shape': [...], u'has': [...]}) at remote 0x7f9927678740>, name=u'Queue', _renamed={}) at remote 0x7f99272b5e50>, client=<SQS(_serializer=<ParamValidationDecorator(_serializer=<QuerySerializer at remote 0x7f992735f6d0>, _param_validator=<ParamValidator at remote 0x7f99273...(truncated)
    s.addHost(hostname=hostname, cluster_user=cluster_user, slots=slots)
#11 Frame 0x2b47dd0, for file /usr/lib/python2.7/site-packages/sqswatcher/sqswatcher.py, line 219, in main (sqsqueue='parallelcluster-c5-18xlarge-ami-02860d917774b37b8-spot-SQS-N8CHAHC5CZHF', table_name='parallelcluster-c5-18xlarge-ami-02860d917774b37b8-spot-DynamoDBTable-1MFDUN1FL9K8G', scheduler='sge', proxy_config=<Config(retries=None, parameter_validation=True, user_agent_extra=None, _user_provided_options={}, s3=None, max_pool_connections=10, signature_version=None, client_cert=None, inject_host_prefix=True, user_agent=None, read_timeout=60, proxies=None, endpoint_discovery_enabled=None, connect_timeout=60, region_name=None) at remote 0x7f9927711c90>, q=<sqs.Queue(meta=<ResourceMeta(service_name='sqs', identifiers=[u'url'], data=None, resource_model=<ResourceModel(_resource_defs=<OrderedDict(_OrderedDict__root=[[[[...], [...], u'Message'], [...], u'Queue'], [...], None], _OrderedDict__map={u'Queue': [...], u'Message': [...]}) at remote 0x7f9927678868>, shape=u'GetQueueAttributesResult', _definition=<OrderedD...(truncated)
    pollQueue(scheduler, q, t, proxy_config)
#14 Frame 0x7f99367a0cc8, for file /usr/bin/sqswatcher, line 11, in <module> ()
    sys.exit(main())
```

so we believe it's stuck here in `/usr/lib/python2.7/site-packages/sqswatcher/plugins/sge.py`:
```
    command = "sudo sh -c \'cd /opt/sge && /opt/sge/inst_sge -noremote -x -auto /opt/parallelcluster/templates/sge/sge_inst.conf\'"
    stdin, stdout, stderr = ssh.exec_command(command)
    while not stdout.channel.exit_status_ready():
        time.sleep(1)
    ssh.close()```
```

which makes sense I think...whatever node it's trying to initialize is long long gone, but this process refuses to keep going because of this `while exist_status_ready` loop

**Comment by lukeseawalker** [comment]
Could you please confirm that you are using "spot" as cluster_type configuration?
Nice catch about the infinite loop, we need to fix this.

**Comment by keien** [comment]
Yes, we are

**Comment by sean-smith** [comment]
Thanks @keien

We’ve marked this as a bug and will update this thread when we have a resolution.

**Comment by keien** [comment]
any updates to this? we've been running into this again recently with p3.2xlarge spot instances during high-demand time periods

**Comment by lukeseawalker** [comment]
it's in the roadmap but we don't have a timeline we can share at this time.

**Comment by demartinofra** [comment]
The latest version of ParallelCluster (2.3.1) already contains some improvements for the above mentioned issue and the sqswatcher daemon should not be stuck on that remote command anymore.
Additional improvements have already been merged into develop and will be released with the next version of ParallelCluster.

What remains to be done to improve overall cluster resiliency is to:
* add some logic to the nodewatcher in order to self terminate a node that does't get added to the scheduler
* auto terminate a node that fails to boot (chef fails to execute recipes)

**Comment by demartinofra** [comment]
This has been addressed in version 2.4.0
